### DISCUSSION TITLE
Newbie needs working multi-threaded Kademlia Example

### DISCUSSION THREAD

**Discussion by mgrand** [libp2p/rust-libp2p#2505] [Q&A] [answered]
The example for kademlia (ipfs-kad) shows how to make a request to Kademlia and then enter an event loop to receive the event produced as a result of the request. I am trying to write an application that initiates Kademlia queries _while_ the event loop is running.

To do this, the calls to initiate Kademlia queries are running in one tokio task and the event loop is running in another. I would like to see a working example that does this.

Here is what I have that does _not_ work. To access the kademlia behaviour to call a `get_record`, I need a mutable reference to the behaviour. To get the mutable reference to the behaviour, I need to call the swarm's behaviour_mut method, which requires a mutable reference to the Swarm. To poll the swarm for events, the event polling thread also need a mutable reference to the swarm. All of the examples I have seen for polling the Swarm involve methods that don't return until there is an event to be returned.

The only way I know for two threads to have a mutable reference to a Swarm is for them to get the reference through a mutex. My problem is that this deadlocks.

The event loop gets the mutable reference from the mutex and polls the Swarm which has no events, so it waits. The other thread wants to get a mutable reference from the mutex, but never gets it because it is held by the event loop that won't relinquish it until it receives and event.

How do I structure something that works?

To frame my question in terms of the example in https://github.com/libp2p/rust-libp2p/blob/master/examples/ipfs-kad.rs, how can I change it so that the event loop is started by calling `spawn` instead of `block_on` and the main task is able to start a second Kademlia query after the event loop is started?

**Comment by demfabris** [comment, marked as answer]
Simply use a `mpsc::channel`. The receiver end will be inside your event loop waiting for instructions. (probably inside a `tokio::select!`) 
The transmitter you take wherever you want

I have a similar example: [gistit-daemon](https://github.com/demfabris/gistit/blob/1a2b4669c7bc3b31f560b276ad13947248c8716b/gistit-daemon/src/node.rs#L106)
But I'm using unix sockets here, where you'd be using just a `mpsc::channel`

Another similar example: [file-sharing](https://github.com/libp2p/rust-libp2p/blob/master/examples/file-sharing.rs)

**Reply by mgrand** [reply]
Thank you.

**Reply by mxinden** [reply]
Thanks @demfabris for stepping in here.
